File: src/main/java/org/mitallast/queue/security/ECDHFlow.java
package org.mitallast.queue.security;

import io.netty.util.AttributeKey;
import org.conscrypt.OpenSSLProvider;

import javax.crypto.Cipher;
import javax.crypto.KeyAgreement;
import javax.crypto.SecretKey;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.lang.reflect.Constructor;
import java.lang.reflect.Field;
import java.lang.reflect.Modifier;
import java.nio.ByteBuffer;
import java.security.*;
import java.security.spec.ECGenParameterSpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

@SuppressWarnings("unchecked")
public class ECDHFlow {

    public static final AttributeKey<ECDHFlow> key = AttributeKey.valueOf("ECDH");

    static {
        String errorString = "Failed manually overriding key-length permissions.";
        int newMaxKeyLength;
        try {
            if ((newMaxKeyLength = Cipher.getMaxAllowedKeyLength("AES")) < 256) {
                Class c = Class.forName("javax.crypto.CryptoAllPermissionCollection");
                Constructor con = c.getDeclaredConstructor();
                con.setAccessible(true);
                Object allPermissionCollection = con.newInstance();
                Field f = c.getDeclaredField("all_allowed");
                f.setAccessible(true);
                f.setBoolean(allPermissionCollection, true);

                c = Class.forName("javax.crypto.CryptoPermissions");

                con = c.getDeclaredConstructor();
                con.setAccessible(true);
                Object allPermissions = con.newInstance();
                f = c.getDeclaredField("perms");
                f.setAccessible(true);

                ((Map) f.get(allPermissions)).put("*", allPermissionCollection);

                c = Class.forName("javax.crypto.JceSecurityManager");
                f = c.getDeclaredField("defaultPolicy");
                f.setAccessible(true);
                Field mf = Field.class.getDeclaredField("modifiers");
                mf.setAccessible(true);
                mf.setInt(f, f.getModifiers() & ~Modifier.FINAL);
                f.set(null, allPermissions);

                newMaxKeyLength = Cipher.getMaxAllowedKeyLength("AES");
            }
        } catch (Exception e) {
            throw new RuntimeException(errorString, e);
        }
        if (newMaxKeyLength < 256)
            throw new RuntimeException(errorString); // hack failed
    }

    private static final String PROVIDER = "GC";
    private static final String ECC_KEY_TYPE = "EC";
    private static final String ECC_CURVE = "secp224r1";
    private static final String ECC_SIGNATURE = "SHA1withECDSA";
    private static final String ECDH_AGREEMENT = "ECDH";
    private static final String AES256 = "AES/CBC/PKCS5Padding";
    private static final String DIGEST = "SHA-256";

    static {
        Security.addProvider(new OpenSSLProvider(PROVIDER));
    }

    private enum State {START, AGREEMENT}

    private final KeyPair keyPair;
    private State state;
    private PublicKey otherPublicKey;
    private SecretKey secretKey;

    private final SecurityService securityService;
    private final Signature signSignature;
    private final Signature verifySignature;
    private final Cipher encryptCipher;
    private final Cipher decryptCipher;
    private final CompletableFuture<Void> agreementFuture;

    public ECDHFlow(SecurityService securityService) throws GeneralSecurityException {
        this.securityService = securityService;
        this.signSignature = Signature.getInstance(ECC_SIGNATURE, PROVIDER);
        this.verifySignature = Signature.getInstance(ECC_SIGNATURE, PROVIDER);
        this.encryptCipher = Cipher.getInstance(AES256);
        this.decryptCipher = Cipher.getInstance(AES256);
        this.agreementFuture = new CompletableFuture<>();

        KeyPairGenerator generator = KeyPairGenerator.getInstance(ECC_KEY_TYPE, PROVIDER);
        ECGenParameterSpec spec = new ECGenParameterSpec(ECC_CURVE);
        generator.initialize(spec);
        keyPair = generator.genKeyPair();

        state = State.START;
    }

    public ECDHRequest requestStart() throws GeneralSecurityException {
        byte[] publicKey = keyPair.getPublic().getEncoded();

        byte[] sign = securityService.sign(publicKey);
        byte[] encrypted = securityService.encrypt(publicKey);
        return new ECDHRequest(sign, encrypted);
    }

    public ECDHResponse responseStart() throws GeneralSecurityException {
        byte[] publicKey = keyPair.getPublic().getEncoded();

        byte[] sign = securityService.sign(publicKey);
        byte[] encrypted = securityService.encrypt(publicKey);
        return new ECDHResponse(sign, encrypted);
    }

    public void keyAgreement(ECDHRequest start) throws Exception {
        byte[] decrypted = securityService.decrypt(start.encodedKey());
        if (!securityService.verify(decrypted, start.sign())) {
            throw new IllegalArgumentException("not verified");
        }
        keyAgreement(decrypted);
    }

    public void keyAgreement(ECDHResponse start) throws Exception {
        byte[] decrypted = securityService.decrypt(start.encodedKey());
        if (!securityService.verify(decrypted, start.sign())) {
            throw new IllegalArgumentException("not verified");
        }
        keyAgreement(decrypted);
    }

    public synchronized void keyAgreement(byte[] publicKey) throws Exception {
        if (state != State.START) {
            throw new IllegalStateException("dh not in started state");
        }
        this.state = State.AGREEMENT;
        KeyFactory kf = KeyFactory.getInstance(ECC_KEY_TYPE, PROVIDER);
        X509EncodedKeySpec pkSpec = new X509EncodedKeySpec(publicKey);
        otherPublicKey = kf.generatePublic(pkSpec);

        // Perform key agreement
        KeyAgreement ka = KeyAgreement.getInstance(ECDH_AGREEMENT);
        ka.init(keyPair.getPrivate());
        ka.doPhase(otherPublicKey, true);
        byte[] sharedSecret = ka.generateSecret();

        // Derive a key from the shared secret and both public keys
        MessageDigest hash = MessageDigest.getInstance(DIGEST);
        hash.update(sharedSecret);
        // Simple deterministic ordering
        List<ByteBuffer> keys = Arrays.asList(
                ByteBuffer.wrap(keyPair.getPublic().getEncoded()),
                ByteBuffer.wrap(publicKey));
        Collections.sort(keys);
        hash.update(keys.get(0));
        hash.update(keys.get(1));

        byte[] derivedKey = hash.digest();
        secretKey = new SecretKeySpec(derivedKey, "AES");

        agreementFuture.complete(null);
    }

    public byte[] sign(byte[] data) throws Exception {
        signSignature.initSign(keyPair.getPrivate());
        signSignature.update(data);
        return signSignature.sign();
    }

    public boolean verify(byte[] data, byte[] sign) throws Exception {
        verifySignature.initVerify(otherPublicKey);
        verifySignature.update(data);
        return verifySignature.verify(sign);
    }

    public ECDHEncrypted encrypt(byte[] data) throws Exception {
        if (state != State.AGREEMENT) {
            throw new IllegalStateException("no shared key");
        }

        encryptCipher.init(Cipher.ENCRYPT_MODE, secretKey);
        AlgorithmParameters params = encryptCipher.getParameters();
        byte[] iv = params.getParameterSpec(IvParameterSpec.class).getIV();
        byte[] encrypted = encryptCipher.doFinal(data);
        byte[] sign = sign(data);
        return new ECDHEncrypted(sign, iv, encrypted);
    }

    public byte[] decrypt(ECDHEncrypted encrypted) throws Exception {
        if (state != State.AGREEMENT) {
            throw new IllegalStateException("no shared key");
        }

        decryptCipher.init(Cipher.DECRYPT_MODE, secretKey, new IvParameterSpec(encrypted.iv()));
        byte[] decrypted = decryptCipher.doFinal(encrypted.encrypted());
        if (!verify(decrypted, encrypted.sign())) {
            throw new IllegalArgumentException("not verified");
        }
        return decrypted;
    }

    public boolean isAgreement() {
        return state == State.AGREEMENT;
    }

    public CompletableFuture<Void> agreementFuture() {
        return agreementFuture;
    }
}

File: src/main/java/org/mitallast/queue/security/ECDHFlow.kt
package org.mitallast.queue.security

import io.netty.util.AttributeKey
import org.conscrypt.OpenSSLProvider
import java.lang.reflect.Field
import java.lang.reflect.Modifier
import java.nio.ByteBuffer
import java.security.*
import java.security.spec.ECGenParameterSpec
import java.security.spec.X509EncodedKeySpec
import java.util.*
import java.util.concurrent.CompletableFuture
import javax.crypto.Cipher
import javax.crypto.KeyAgreement
import javax.crypto.SecretKey
import javax.crypto.spec.SecretKeySpec

class ECDHFlow constructor(private val securityService: SecurityService) {

    private enum class State {
        START, AGREEMENT
    }

    private val keyPair: KeyPair
    private var state: State = State.START
    private var otherPublicKey: PublicKey? = null
    private var secretKey: SecretKey? = null
    private val agreementFuture = CompletableFuture<Void>()

    init {
        val generator = KeyPairGenerator.getInstance(ECC_KEY_TYPE, PROVIDER)
        val spec = ECGenParameterSpec(ECC_CURVE)
        generator.initialize(spec)
        keyPair = generator.genKeyPair()
    }

    fun requestStart(): ECDHRequest {
        val publicKey = keyPair.public.encoded

        val sign = securityService.sign(publicKey)
        val encrypted = securityService.encrypt(publicKey)
        return ECDHRequest(sign, encrypted)
    }

    fun responseStart(): ECDHResponse {
        val publicKey = keyPair.public.encoded

        val sign = securityService.sign(publicKey)
        val encrypted = securityService.encrypt(publicKey)
        return ECDHResponse(sign, encrypted)
    }

    fun keyAgreement(start: ECDHRequest) {
        val decrypted = securityService.decrypt(start.encodedKey)
        if (!securityService.verify(decrypted, start.sign)) {
            throw IllegalArgumentException("not verified")
        }
        keyAgreement(decrypted)
    }

    fun keyAgreement(start: ECDHResponse) {
        val decrypted = securityService.decrypt(start.encodedKey)
        if (!securityService.verify(decrypted, start.sign)) {
            throw IllegalArgumentException("not verified")
        }
        keyAgreement(decrypted)
    }

    private fun keyAgreement(publicKey: ByteArray) {
        if (state != State.START) {
            throw IllegalStateException("dh not in started state")
        }
        this.state = State.AGREEMENT
        val kf = KeyFactory.getInstance(ECC_KEY_TYPE, PROVIDER)
        val pkSpec = X509EncodedKeySpec(publicKey)
        otherPublicKey = kf.generatePublic(pkSpec)

        // Perform key agreement
        val ka = KeyAgreement.getInstance(ECDH_AGREEMENT)
        ka.init(keyPair.private)
        ka.doPhase(otherPublicKey, true)
        val sharedSecret = ka.generateSecret()

        // Derive a key from the shared secret and both public keys
        val hash = MessageDigest.getInstance(DIGEST)
        hash.update(sharedSecret)
        // Simple deterministic ordering
        val keys = Arrays.asList(
            ByteBuffer.wrap(keyPair.public.encoded),
            ByteBuffer.wrap(publicKey))
        keys.sort()
        hash.update(keys[0])
        hash.update(keys[1])

        val derivedKey = hash.digest()
        secretKey = SecretKeySpec(derivedKey, "AES")

        agreementFuture.complete(null)
    }

    fun secretKey(): SecretKey {
        if (state != State.AGREEMENT) {
            throw IllegalStateException("no shared key")
        }
        return secretKey!!
    }

    val isAgreement: Boolean
        get() = state == State.AGREEMENT

    fun agreementFuture(): CompletableFuture<Void> {
        return agreementFuture
    }

    companion object {
        val key: AttributeKey<ECDHFlow> = AttributeKey.valueOf("ECDH")

        init {
            try {
                val newMaxKeyLength = Cipher.getMaxAllowedKeyLength("AES")
                if (newMaxKeyLength < 256) {
                    var c = Class.forName("javax.crypto.CryptoAllPermissionCollection")
                    var con = c.getDeclaredConstructor()
                    con.isAccessible = true
                    val allPermissionCollection = con.newInstance()
                    var f = c.getDeclaredField("all_allowed")
                    f.isAccessible = true
                    f.setBoolean(allPermissionCollection, true)

                    c = Class.forName("javax.crypto.CryptoPermissions")

                    con = c.getDeclaredConstructor()
                    con.isAccessible = true
                    val allPermissions = con.newInstance()
                    f = c.getDeclaredField("perms")
                    f.isAccessible = true

                    @Suppress("UNCHECKED_CAST")
                    (f.get(allPermissions) as MutableMap<String, Any>).put("*", allPermissionCollection)

                    c = Class.forName("javax.crypto.JceSecurityManager")
                    f = c.getDeclaredField("defaultPolicy")
                    f.isAccessible = true
                    val mf = Field::class.java.getDeclaredField("modifiers")
                    mf.isAccessible = true
                    mf.setInt(f, f.modifiers and Modifier.FINAL.inv())
                    f.set(null, allPermissions)
                }
            } catch (e: Exception) {
                throw RuntimeException("Failed manually overriding key-length permissions.", e)
            }

            if (Cipher.getMaxAllowedKeyLength("AES") < 256)
                throw RuntimeException("Failed manually overriding key-length permissions.") // hack failed
        }

        const val PROVIDER = "GC"
        const val ECC_KEY_TYPE = "EC"
        const val ECC_CURVE = "secp224r1"
        const val ECDH_AGREEMENT = "ECDH"
        const val AES256 = "AES/CBC/PKCS5Padding"
        const val DIGEST = "SHA-256"
        const val HmacSHA256 = "HmacSHA256"

        init {
            Security.addProvider(OpenSSLProvider(PROVIDER))
        }
    }
}
